pandas/io/formats/xml.py

```diff
@@ -0,0 +1,644 @@
+        """
+
+        raise AbstractMethodError(self)
+
+    def other_namespaces(self) -> dict:
+        """
+        Define other namespaces.
+
+        This method will build dictionary of namespaces attributes
+        for root element, conditionally with optional namespaces and
+        prefix.
+        """
+
+        nmsp_dict: Dict[str, str] = {}
+        if self.namespaces and self.prefix is None:
+            nmsp_dict = {"xmlns": n for p, n in self.namespaces.items() if p != ""}
+
+        if self.namespaces and self.prefix:
+            nmsp_dict = {"xmlns": n for p, n in self.namespaces.items() if p == ""}
+
+        return nmsp_dict
+
+    def build_attribs(self) -> None:
+        """
+        Create attributes of row.
+
+        This method adds attributes using attr_cols to row element and
+        works with tuples for multindex or hierarchical columns.
+        """
+
+        raise AbstractMethodError(self)
+
+    def build_elems(self) -> None:
+        """
+        Create child elements of row.
+
+        This method adds child elements using elem_cols to row element and
+        works with tuples for multindex or hierarchical columns.
+        """
+
+        raise AbstractMethodError(self)
+
+    def _get_data_from_filepath(self, filepath_or_buffer):
+        """
+        Extract raw XML data.
+
+        The method accepts three input types:
+            1. filepath (string-like)
+            2. file-like object (e.g. open file object, StringIO)
+            3. XML string or bytes
+
+        This method turns (1) into (2) to simplify the rest of the processing.
+        It returns input types (2) and (3) unchanged.
+        """
+        filepath_or_buffer = stringify_path(filepath_or_buffer)
+
+        if (
+            isinstance(filepath_or_buffer, str)
+            and not filepath_or_buffer.startswith(("<?xml", "<"))
+        ) and (
+            not isinstance(filepath_or_buffer, str)
+            or is_url(filepath_or_buffer)
+            or is_fsspec_url(filepath_or_buffer)
+            or file_exists(filepath_or_buffer)
+        ):
+            with get_handle(
+                filepath_or_buffer,
+                "r",
+                encoding=self.encoding,
+                compression=self.compression,
+                storage_options=self.storage_options,
+            ) as handle_obj:
+                filepath_or_buffer = (
+                    handle_obj.handle.read()
+                    if hasattr(handle_obj.handle, "read")
+                    else handle_obj.handle
+                )
+
+        return filepath_or_buffer
+
+    def _preprocess_data(self, data):
+        """
+        Convert extracted raw data.
+
+        This method will return underlying data of extracted XML content.
+        The data either has a `read` attribute (e.g. a file object or a
+        StringIO/BytesIO) or is a string or bytes that is an XML document.
+        """
+        if isinstance(data, str):
```

Comment:
hmm why do we accept bytes here? (and not just string)?

Reply:
why is this a method on the class?

Reply:
Both `_get_data_from_filepath` and `_preprocess_data` methods were borrowed from pandas.io.json._json inside the `JsonReader` class. How to adjust here?

Reply:
oh, ok yeah just make a module level function if you need to do this then